lightningd/peer_control.c

```diff
@@ -2045,6 +2045,46 @@ static int peer_start_closingd_after_shutdown(struct peer *peer, const u8 *msg,
 	return -1;
 }
 
+/**
+ * transactional_msg_handler - Wrap a msg handler in a DB transaction
+ *
+ * Initiates a DB transaction, executes the @handler with the given
+ * arguments and terminates the transaction after the handler
+ * returns. If the handler's return value is negative the transaction
+ * is aborted. If the handler returns a positive value we attempt to
+ * commit the transaction. Returns the return value from the handler
+ * or -1 on commit failure.
+ */
```

Comment:
Yep, this patch is a real win.  I'm not sure where the nested transacitons are coming from though.. I'll have to apply these last two patches and see what crashes?

Reply:
The nesting mainly comes from the fact that I'd like to be able to ensure transactional behavior in lower components independently from the call-path. For example `wallet_channel_save` uses a series of tightly coupled `INSERT` and `UPDATE` statements that must commit atomically, no matter how we got there or whether the caller took care of starting a transaction, hence the nesting.